lib/srv/db/mongodb/connect.go

```diff
@@ -153,6 +161,43 @@ func (e *Engine) getConnectionOptions(ctx context.Context, sessionCtx *common.Se
 	}, nil
 }
 
+func (e *Engine) getAuthenticator(ctx context.Context, sessionCtx *common.Session) (authenticator auth.Authenticator, err error) {
+	// Currently, the MongoDB Atlas IAM Authentication doesn't work with IAM
+	// users. Here we provide a better error message to the users.
+	if sessionCtx.Database.GetType() == types.DatabaseTypeMongoAtlas && awsutils.IsUserARN(sessionCtx.DatabaseUser) {
+		return nil, trace.BadParameter("MongoDB Atlas AWS IAM Authentication with IAM users is not supported.")
+	}
+
+	switch {
+	case sessionCtx.Database.GetType() == types.DatabaseTypeMongoAtlas && awsutils.IsRoleARN(sessionCtx.DatabaseUser):
+		e.Log.Debug("Authenticating to database using AWS IAM authentication")
+		username, password, sessToken, authErr := e.Auth.GetAtlasIAMToken(ctx, sessionCtx)
+		if err != nil {
+			return nil, trace.Wrap(authErr)
+		}
```

Comment:
seems that error is assigned to `authErr` but the `if check`  is based on `err` variable: 
```suggestion
		username, password, sessToken, err := e.Auth.GetAtlasIAMToken(ctx, sessionCtx)
		if err != nil {
			return nil, trace.Wrap(err)
		}
```